### PR TITLE
DAOS-9358 control: fix server config nr_hugepages test variants

### PR DESCRIPTION
**Pull request by tanabarr**
Add range validation for nr_hugepages, create faults and adjust tests.

Features: control

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

### CODE REVIEW

**Review by kjacque — APPROVED** [review]

File: src/control/server/config/server.go [hunk 1]
```diff
@@ -528,11 +533,14 @@ func (cfg *Server) Validate(log logging.Logger, hugePageSize int, fis *hardware.
 
 		// If the config doesn't specify hugepages, use the minimum. Otherwise, validate
 		// that the configured amount is sufficient.
-		if cfg.NrHugepages == 0 {
+		switch {
+		case cfg.NrHugepages == -1:
+			return FaultConfigHugepagesDisabled
+		case cfg.NrHugepages == 0:
```
Comment (kccain) [inline src/control/server/config/server.go:539]: does this case correspond to both nr_hugepages: 0 specified explicitly in the daos_server.yml file, as well as no nr_hugepages present in the config file?

File: src/control/server/config/server_test.go [hunk 2]
```diff
@@ -470,7 +471,82 @@ func TestServerConfig_Validation(t *testing.T) {
...
+									WithScmMountPoint("/foo"),
+							).
+							WithFabricInterfacePort(1234).
+							WithFabricInterface("eth0").
+							WithTargetCount(8).
+							WithPinnedNumaNode(0),
+					)
+			},
+		},
+		"insufficient hugepages set in config": {
```
Comment (kccain) [inline src/control/server/config/server_test.go:549]: would it be worth a test case like this that has no bdevs in the configuration? and the expectation (I guess) is that the configuration would be valid / no expected error?

File: src/control/server/config/server_test.go [hunk 3]
```diff
@@ -533,6 +609,49 @@ func TestServerConfig_Validation(t *testing.T) {
 						WithPinnedNumaNode(0),
 				),
 		},
+		"calculated hugepages; emulated bdevs configured": {
```
Comment (kccain) [inline src/control/server/config/server_test.go:612]: Does "calculated" huge pages refer to no nr_hugepages in the server config?

is it worth a test case for no nr_hugepages specified, with no bdevs in the configuration? Not sure what the expConfig NrHugePages would be though.

File: src/tests/ftest/server/daos_server_config.yaml [hunk 4]
```diff
@@ -93,10 +93,20 @@ server_config_val: !mux
       - "nr_hugepages"
       - 1000000000000000000
       - "FAIL"
-  nr_hugepages_negative_int:
+  nr_hugepages_large_negative_int:
     config_val:
       - "nr_hugepages"
       - -100
+      - "FAIL"
+  nr_hugepages_small_negative_int:
```
Comment (kccain) [inline src/tests/ftest/server/daos_server_config.yaml:101]: maybe this is a question for a test developer @phender - would the name of this case better be "hugepages_disabled"?

Also with this test are there bdevs in the configuration or not?

**Review by kccain — COMMENTED** [review]
Just a quick pass with some questions.

File: src/control/server/config/server.go (hunk 1 above)
Comment (tanabarr) [reply src/control/server/config/server.go:539]: yes, zero-value of field and explicit passing of 0

File: src/control/server/config/server_test.go (hunk 3 above)
Comment (tanabarr) [reply src/control/server/config/server_test.go:612]: renamed the test text for clarity and added test case for none specified with no bdevs configured

File: src/tests/ftest/server/daos_server_config.yaml (hunk 4 above)
Comment (tanabarr) [reply src/tests/ftest/server/daos_server_config.yaml:101]: I think that would be a better name, however I'm going to remove this added test case for the following reason.

The result of disabling hugepages is not deterministic in this case, although the validation phase during daos_server can be determined and would indeed pass for `nr_hugepages: -1`, subsequent calls to commands that use the SPDK bindings may fail depending on system environment (the daos_server_config test performs storage format and scan). 
Some examples:
- If the system has NVMe SSDs currently bound to VFIO but no hugepages have been allocated then we will see permission denied errors from storage scan.
- If the system has NVMe SSDs currently bound to VFIO by a different user and hugepages have been allocated then we will probably see VFIO device in use errors from storage scan.
- If SSDs are bound to the kernel driver then storage scan will complete successfully.

Disabling hugepages is a documented edge case and as such I believe we have sufficient test coverage in the control plane unit tests that verify the setting behaviour.

Comment (tanabarr) [reply src/tests/ftest/server/daos_server_config.yaml:101]: bdevs are not configured in this test

File: src/control/server/config/server_test.go [hunk 5]
```diff
@@ -568,7 +568,64 @@ func TestServerConfig_Validation(t *testing.T) {
...
+								WithScmRamdiskSize(1).
+								WithScmMountPoint("/foo"),
+						).
+						WithFabricInterfacePort(1234).
+						WithFabricInterface("eth0").
+						WithTargetCount(8).
+						WithPinnedNumaNode(0),
+				),
+		},
+		"zero hugepages set in config": {
```
Comment (kccain) [inline src/control/server/config/server_test.go:628]: very minor, may not be worth it if the rest of the PR has no comments. Consider adding "; bdevs configured" to the name

**Review by mjmac — DISMISSED** [review]
The latest commit in the stack must have the following pragma (it's not sufficient to just put it in the first commit or update the gh message):
`Features: control`

**Review by kjacque — APPROVED** [review]

**Review by kccain — APPROVED** [review]

**Pull request merged**